streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/GraphGraceSearchUtil.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.kstream.internals.graph;
+
+import org.apache.kafka.streams.errors.TopologyException;
+import org.apache.kafka.streams.kstream.SessionWindows;
+import org.apache.kafka.streams.kstream.Windows;
+import org.apache.kafka.streams.kstream.internals.KStreamSessionWindowAggregate;
+import org.apache.kafka.streams.kstream.internals.KStreamWindowAggregate;
+import org.apache.kafka.streams.processor.ProcessorSupplier;
+
+public final class GraphGraceSearchUtil {
+    private GraphGraceSearchUtil() {}
+
+    public static long findAndVerifyWindowGrace(final StreamsGraphNode streamsGraphNode) {
+        return findAndVerifyWindowGrace(streamsGraphNode, "");
+    }
+
+    private static long findAndVerifyWindowGrace(final StreamsGraphNode streamsGraphNode, final String chain) {
+        // error base case: we traversed off the end of the graph without finding a window definition
+        if (streamsGraphNode == null) {
+            throw new TopologyException(
+                "Window close time is only defined for windowed computations. Got [" + chain + "]."
+            );
+        }
+        // base case: return if this node defines a grace period.
+        {
+            final Long gracePeriod = extractGracePeriod(streamsGraphNode);
+            if (gracePeriod != null) {
+                return gracePeriod;
+            }
+        }
+
+        final String newChain = chain.equals("") ? streamsGraphNode.nodeName() : streamsGraphNode.nodeName() + "->" + chain;
+
+        if (streamsGraphNode.parentNodes().isEmpty()) {
+            // error base case: we traversed to the end of the graph without finding a window definition
+            throw new TopologyException(
+                "Window close time is only defined for windowed computations. Got [" + newChain + "]."
+            );
+        }
+
+        // recursive case: all parents must define a grace period, and they must match.
```

Comment:
The last point is correct. I thought that we were already aware of this shortcoming... but now I don't remember who I was talking to about it.

About the parents... The immediate parent of suppress is a single KTable, but may not be the one with a defined grace period. It might be a filter, in which case we need to examine the parent of the filter, or it might be a join, in which case we need to examine *both* parents of the join.

The fact that we currently have a design flaw that prevents this situation doesn't imply that we should encode this limitation here. Once we fix that design flaw, we would have to remember that we also coded that flaw into this method and come back to revert it to the state it's in right now!

Reply:
Hmmm... I did originally not consider KTable-KTable join -- however, a grace period is only defined for windowed aggregations, right? Thus, this would only make sense if two windowed-KTables are joined?

Thus, if we consider that we might fix joining two Windowed-KTables in the future (was is broker atm), I am wondering, if we should use the maximum grace period over both base-join-tables as required bound for the suppression, instead of forcing both base-windowed-tables to have the same grace period configured?

Reply:
Yeah, we could make it more permissive that way.

This is a discussion for the future when we do fix that operation, but it seems safest to support that join only when both streams have the exact same window configuration (otherwise there's no guarantee that the streams have any keys in common).

In such a situation, we wouldn't have to worry about enforcing it here. But for now, I was thinking to be strict about the common grace periods as a basic precaution against mixing window types. (even though the grace period isn't part of the windowed key).

Reply:
> otherwise there's no guarantee that the streams have any keys in common

Cannot follow here. Can you elaborate?

Also, thinking about this once more: why do we need to force `suppress` to have a larger grace period that it's parents?

Reply:
What I meant is that if the left stream and the right stream have completely different window specs, then the join will be completely disjoint. This is probably a programming mistake, and I think it's better to fail fast.

Reply:
> have completely different window specs

This makes sense -- however, if no suppress operation is defined, this would not be detected -- and if we put a check somewhere else, suppress does not need to check it either. Does it? Additionally, even if suppress does this check, I would exclude the grace period from the check.

> We set suppress's suppression time equal to the grace period of the parent.

From my understanding, this method should checks that the grace period is larger than the window-size of the upstream KTable -- we don't need to set the grace-period, but just take whatever the users specified for it. There is no comparison of parent grace period and suppress grace period? And there is no need? Or do I miss something?

Reply:
Yeah, I think we're miscommunicating.
Suppression does not have a "grace period", it only has the "emit after" config.

The method `findAndVerifyWindowGrace` only extracts the exact specified grace period, as configured upstream.

We pass the value we get back to `org.apache.kafka.streams.kstream.internals.suppress.SuppressedImpl#buildFinalResultsSuppression`, which then creates a `Suppressed` configuration using that time (the extracted grace period) as the "emit after" config, along with setting the `timeDefinition` to the window end time.

Together these configs cause suppression to emit immediately right at the end of the grace period (as configured somewhere upstream).

Reply:
You raised a separate question about whether we should be strict or permissive if the suppression node actually has multiple parents which specify two different grace periods.

I favored strict because I think this makes the situation more debuggable and comprehensible. But you are correct in stating that it's not a correctness issue, since the grace period doesn't affect the key. For this reason, I would be ok with logging a warning and just using the larger grace period.

The consideration we need to weigh is in that situation with two different windowed tables getting merged/joined, assuming they only differ in grace period, how can we ensure that all operators follow the same strategy of taking the larger grace period. It affects windowed store retention as well as suppression. I think taking the larger is a reasonably obvious choice, so maybe it's not a big deal, though.

Reply:
Thanks for clarification!

Would be good to get input from @guozhangwang @bbejeck about this. Should we fail? Or should we pick "max"? I still favor "max", but it's not a deal breaker if we fail instead.

Nevertheless, I don't think `suppress` should check the window spec (size/advance etc) -- if a user does a "weird" join and we want to disallow it, this check should be done in `join()` instead.

Reply:
I would favor taking the `max` as well, as, IMHO, we can't account for all use cases so better to chose max.  We should document the behavior so users can be made aware of what happens with setting different grace periods

Would it be too much to make this a configurable item?  I'm not sure as we have several config items to consider already.

Reply:
Regarding the `.filter(..., Materialize)` issue, yes this is a known bug. I think @abbccdda has raised a KIP for fixing this.

Regarding the windowed-KTable / KTable join, there are some very old discussions before on how to tackle it (https://issues.apache.org/jira/browse/KAFKA-7368 is filed recently), and the idea at that time was:

1) we would require the joining table's window-spec to be well aligned, i.e. they must have the same length. Note since window boundaries are the same as they are all starting from the epoch time, it means that same window length guarantees aligned windows, and each the join operation becoming joining each paired windows of the table. We do not yet have session windows at all so this was not discussed, but with session windows it is definitely more complicated..

As for grace period, I think we do not need to make it strict that requires grace to be the same as well. Personally I think either `min` or `max` are fine (I'm slightly leaning towards `min` though :P).

Reply:
Why `min`? `max` seems to be inclusive and guarantees to respect the configs of both upstream operators, while `min` does not?

Reply:
Ok, 
* fwiw: I 100% agree that it's not this component's job to verify other aspects of the window spec. It should only care about grace period.
* I think the semantics are perfectly well defined with `max`, and I buy these arguments that it's unnecessary to fail. I'd like to log a warning (just during the topology build) in case the mis-match was accidental.

@guozhangwang :
I agree with @mjsax , I don't think that `min` has the right semantics. The purpose of configuring the suppression interval equal to the grace period to begin with is that we already know that the window results will never be updated after the grace period ends. If we set the suppression smaller than the grace period (or in this case _one_ of the grace periods), then there will be an inconsistency between the aggregation results upstream vs. downstream of suppression.

In fact, it's generally ok if we make the suppression interval _larger_ than the grace period. It just means that we'll emit the final result "after" the window closes, not "at" the window close.

Thus, `max` satisfies everyone's semantics and ensures consistent results throughout the topology. The "more graceful" parent will see its final results close after its grace period expires, and the "less graceful" one has to wait longer, but they will both never see their materialized state differ from the results downstream of the suppression.

Reply:
From my understanding, using max will be the default behavior -- users can still manually specify a smaller one, right? It must just be larger than window-size?

Reply:
Ok, recall that we are talking about the special case in which the suppress operation has two parents who have different grace periods configured. In this case, "max" means that the suppression will be configured with the larger of the two parents' grace periods.

Normally, suppression only has one parent, in which case, it's configured to suppress for N ms, where N equals it's parent's grace period. N might be 0.

If you select "final results", there is *no* option to configure the suppression interval. It is _always_ taken from the parent(s)'s configured grace period. To do anything else would threaten consistency.

However, it's always possible just to do regular intermediate suppression, and choose any time you like, larger or smaller than the window size. You're just not guaranteed to get exactly one result per key/window if you pick any time shorter than window size + grace period.